File: bench.js
#!/usr/bin/env node
/*
 * bench.js -- the JavaScript entry in the language speed comparison.
 *
 * Usage: node bench.js <benchmark> <size> [reps]
 * Prints: OK <benchmark> <checksum> <compute_milliseconds>
 *
 * Same algorithm, same deterministic input, same checksum as every other
 * bench.* in this suite.
 *
 * Plain JavaScript on the default V8 configuration -- no WASM, no native
 * addons, no flags. Typed arrays are used where the other languages use a
 * flat array of a fixed width, because that is the honest equivalent: a
 * plain JS array of numbers is a different data structure, not a slower one.
 */
'use strict';

// ---------- shared deterministic PRNG (identical in every language here) ----------
// The others do `state = state * 6364136223846793005 + 1442695040888963407`
// in one 64-bit multiply. JavaScript has no 64-bit integer that is fast --
// BigInt would be correct and roughly an order of magnitude slower -- so the
// multiply is done by hand in 32-bit halves. This is the price of admission
// for JS the way the hand-rolled hash table is the price of admission for C.
const MUL_HI = 0x5851f42d, MUL_LO = 0x4c957f2d;   // 6364136223846793005
const ADD_HI = 0x14057b7e, ADD_LO = 0xf767814f;   // 1442695040888963407

let stateHi = 0, stateLo = 0;

function rngSeed(s) {
  stateHi = Math.floor(s / 4294967296) >>> 0;
  stateLo = (s >>> 0);
}

function rngNext() {
  const aLo = stateLo, aHi = stateHi;
  // 32x32 -> 64 of the low halves, via 16-bit limbs
  const a0 = aLo & 0xffff, a1 = aLo >>> 16;
  const b0 = MUL_LO & 0xffff, b1 = MUL_LO >>> 16;
  const p0 = a0 * b0;
  const p1 = a1 * b0 + (p0 >>> 16);
  let carry = p1 >>> 16;
  const p1b = (p1 & 0xffff) + a0 * b1;
  carry += p1b >>> 16;
  const mulLo = (((p1b & 0xffff) << 16) | (p0 & 0xffff)) >>> 0;
  // high half: carry + a1*b1 + the two cross terms that only affect bits 32+
  const mulHi = (carry + a1 * b1 + Math.imul(aLo, MUL_HI) + Math.imul(aHi, MUL_LO)) >>> 0;

  const lo = (mulLo + ADD_LO) >>> 0;
  stateLo = lo;
  stateHi = (mulHi + ADD_HI + (lo < mulLo ? 1 : 0)) >>> 0;
  return stateHi >>> 1;   // (state >> 33) is just the high word shifted once
}

// ---------- 1. mandelbrot: tight floating-point loop, zero allocation ----------
function benchMandelbrot(n) {
  let total = 0;
  for (let py = 0; py < n; py++) {
    const ci = 2.0 * py / n - 1.0;
    for (let px = 0; px < n; px++) {
      const cr = 2.0 * px / n - 1.5;
      let zr = 0.0, zi = 0.0;
      let i = 0;
      while (i < 255) {
        const zr2 = zr * zr, zi2 = zi * zi;
        if (zr2 + zi2 > 4.0) break;
        zi = 2.0 * zr * zi + ci;
        zr = zr2 - zi2 + cr;
        i++;
      }
      total += i;
    }
  }
  return total;
}

// ---------- 2. sieve of eratosthenes: integer math over a large flat array ----------
function benchSieve(n) {
  const comp = new Uint8Array(n + 1);
  let count = 0;
  for (let i = 2; i <= n; i++) {
    if (comp[i] === 0) {
      count++;
      for (let j = i * i; j <= n; j += i) comp[j] = 1;
    }
  }
  return count;
}

// ---------- 3. quicksort: branches, swaps, recursion, random memory access ----------
// Hand-written, not Array.prototype.sort -- the point is to measure the
// language, not the quality of its sort library.
function insertionSort(a, lo, hi) {
  for (let i = lo + 1; i <= hi; i++) {
    const v = a[i];
    let j = i - 1;
    while (j >= lo && a[j] > v) { a[j + 1] = a[j]; j--; }
    a[j + 1] = v;
  }
}

function quicksort(a, lo, hi) {
  while (hi - lo > 16) {
    const mid = lo + ((hi - lo) >> 1);
    let t;
    // median-of-three: order a[lo] <= a[mid] <= a[hi]
    if (a[mid] < a[lo])  { t = a[mid]; a[mid] = a[lo];  a[lo]  = t; }
    if (a[hi]  < a[lo])  { t = a[hi];  a[hi]  = a[lo];  a[lo]  = t; }
    if (a[hi]  < a[mid]) { t = a[hi];  a[hi]  = a[mid]; a[mid] = t; }
    const pivot = a[mid];

    let i = lo, j = hi;
    while (i <= j) {
      while (a[i] < pivot) i++;
      while (a[j] > pivot) j--;
      if (i <= j) { t = a[i]; a[i] = a[j]; a[j] = t; i++; j--; }
    }
    // recurse into the smaller half, loop on the larger: depth stays O(log n)
    if (j - lo < hi - i) { quicksort(a, lo, j); lo = i; }
    else                 { quicksort(a, i, hi); hi = j; }
  }
  insertionSort(a, lo, hi);
}

function benchQuicksort(n) {
  const a = new Uint32Array(n);
  rngSeed(12345);
  for (let i = 0; i < n; i++) a[i] = rngNext();
  quicksort(a, 0, n - 1);
  let h = 0;
  for (let i = 0; i < n; i++) h = (Math.imul(h, 31) + a[i]) >>> 0;  // order-sensitive
  return h;
}

// ---------- 4. word count: string hashing + hash map ----------
// A Map with string keys, one line, like Go's and Java's -- and like theirs,
// the hashing underneath is native code, so this is the benchmark where an
// interpreted-ish language is closest to C.
const VOCAB = 5000;

function benchWordcount(n) {
  rngSeed(12345);
  const words = new Array(VOCAB);
  for (let i = 0; i < VOCAB; i++) {
    const len = 3 + rngNext() % 6;
    let w = '';
    for (let c = 0; c < len; c++) w += String.fromCharCode(97 + rngNext() % 26);
    words[i] = w;
  }

  const counts = new Map();
  let maxc = 0;
  for (let k = 0; k < n; k++) {
    const ra = rngNext() % VOCAB;
    const rb = rngNext() % VOCAB;
    const w = words[(ra * rb) / VOCAB | 0];   // triangular, so counts vary
    const c = (counts.get(w) || 0) + 1;
    counts.set(w, c);
    if (c > maxc) maxc = c;
  }
  return counts.size * 1000003 + maxc;
}

// ---------- 5. binary trees: allocation, pointer chasing, garbage collection ----------
// Nothing here computes; the entire cost is allocating little objects and
// letting V8's generational GC take them back.
function makeTree(d) {
  if (d === 0) return { l: null, r: null };
  return { l: makeTree(d - 1), r: makeTree(d - 1) };
}

function checkTree(t) {
  if (t.l === null) return 1;
  return 1 + checkTree(t.l) + checkTree(t.r);
}

function benchBinarytrees(n) {
  let h = 0;
  for (let k = 0; k < n; k++) {
    const t = makeTree(11);   // 4,095 nodes, built and thrown away
    h = (Math.imul(h, 31) + checkTree(t) + k) >>> 0;
  }
  return h;
}

// ---------- 6. matmul: triple-nested loops over flat 2D arrays ----------
// The matrix entries are < 100, so every product and every row sum fits
// exactly in a double -- no Math.imul needed until the final checksum.
function benchMatmul(n) {
  const a = new Uint32Array(n * n);
  const b = new Uint32Array(n * n);
  const c = new Uint32Array(n * n);
  rngSeed(12345);
  for (let i = 0; i < n * n; i++) a[i] = rngNext() % 100;
  for (let i = 0; i < n * n; i++) b[i] = rngNext() % 100;
  for (let i = 0; i < n; i++) {
    const ib = i * n;
    for (let j = 0; j < n; j++) {
      let s = 0;
      for (let k = 0; k < n; k++) s += a[ib + k] * b[k * n + j];
      c[ib + j] = s;
    }
  }
  let h = 0;
  for (let i = 0; i < n * n; i++) h = (Math.imul(h, 31) + c[i]) >>> 0;
  return h;
}

const BENCHMARKS = {
  mandelbrot: benchMandelbrot,
  sieve: benchSieve,
  quicksort: benchQuicksort,
  wordcount: benchWordcount,
  binarytrees: benchBinarytrees,
  matmul: benchMatmul,
};

function main() {
  const argv = process.argv.slice(2);
  if (argv.length < 2) {
    process.stderr.write('usage: node bench.js <benchmark> <size> [reps]\n');
    return 2;
  }
  const name = argv[0];
  const size = parseInt(argv[1], 10);
  let reps = argv.length > 2 ? parseInt(argv[2], 10) : 1;
  if (!(reps >= 1)) reps = 1;

  const fn = BENCHMARKS[name];
  if (!fn) {
    process.stderr.write('unknown benchmark: ' + name + '\n');
    return 2;
  }

  // Best-of-N: the fastest run is the one least polluted by scheduler noise,
  // cold caches and (for the JIT languages) not-yet-compiled code.
  let best = Infinity;
  let result = null;
  for (let r = 0; r < reps; r++) {
    const t0 = process.hrtime.bigint();
    const v = fn(size);
    const elapsed = Number(process.hrtime.bigint() - t0) / 1e6;
    if (elapsed < best) best = elapsed;
    if (r === 0) result = v;
    else if (v !== result) {
      process.stderr.write('nondeterministic result!\n');
      return 3;
    }
  }

  process.stdout.write('OK ' + name + ' ' + result + ' ' + best.toFixed(3) + '\n');
  return 0;
}

process.exit(main());

File: bench.rs
// bench.rs -- the Rust entry in the language speed comparison.
//
// Usage: bench <benchmark> <size> [reps]
// Prints: OK <benchmark> <checksum> <compute_milliseconds>
//
// Same algorithm, same deterministic input, same checksum as every other
// bench.* in this suite.
//
// Built with a bare `rustc -O` -- no Cargo, no crates, no `unsafe`. Every
// index below is bounds-checked, which is exactly the trade this entry is
// here to price: safety that you are told costs nothing, measured.

use std::collections::HashMap;
use std::env;
use std::process;
use std::time::Instant;

// ---------- shared deterministic PRNG (identical in every language here) ----------
struct Rng(u64);

impl Rng {
    fn new(seed: u64) -> Rng {
        Rng(seed)
    }
    #[inline]
    fn next(&mut self) -> u32 {
        // wrapping_* because overflow panics in debug builds; the other
        // languages wrap silently and we must produce the same bits.
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 33) as u32 // top 31 bits
    }
}

// ---------- 1. mandelbrot: tight floating-point loop, zero allocation ----------
fn bench_mandelbrot(n: i32) -> i64 {
    let mut total: i64 = 0;
    let fnn = n as f64;
    for py in 0..n {
        let ci = 2.0 * py as f64 / fnn - 1.0;
        for px in 0..n {
            let cr = 2.0 * px as f64 / fnn - 1.5;
            let (mut zr, mut zi) = (0.0f64, 0.0f64);
            let mut i = 0;
            while i < 255 {
                let (zr2, zi2) = (zr * zr, zi * zi);
                if zr2 + zi2 > 4.0 {
                    break;
                }
                zi = 2.0 * zr * zi + ci;
                zr = zr2 - zi2 + cr;
                i += 1;
            }
            total += i as i64;
        }
    }
    total
}

// ---------- 2. sieve of eratosthenes: integer math over a large flat array ----------
fn bench_sieve(n: i32) -> i64 {
    let n = n as usize;
    let mut comp = vec![0u8; n + 1];
    let mut count: i64 = 0;
    for i in 2..=n {
        if comp[i] == 0 {
            count += 1;
            let mut j = i * i;
            while j <= n {
                comp[j] = 1;
                j += i;
            }
        }
    }
    count
}

// ---------- 3. quicksort: branches, swaps, recursion, random memory access ----------
// Hand-written, not slice::sort_unstable -- the point is to measure the
// language, not the quality of its sort library.
fn insertion_sort(a: &mut [u32], lo: isize, hi: isize) {
    // Work on the subslice so the bound the optimiser sees is s.len(), not
    // "whatever lo and hi happen to be" -- that folds the per-shift range
    // checks. hi can legitimately be lo-1 (an empty partition), which is why
    // the end is hi+1 rather than an inclusive ..=hi.
    let s = &mut a[lo as usize..(hi + 1) as usize];
    for i in 1..s.len() {
        let v = s[i];
        let mut j = i;
        while j > 0 && s[j - 1] > v {
            s[j] = s[j - 1];
            j -= 1;
        }
        s[j] = v;
    }
}

// Indices are isize rather than usize on purpose: the C original lets `j` run
// one step past `lo`, and a usize would panic there instead of wrapping. Same
// algorithm, same swap sequence, same checksum.
fn quicksort(a: &mut [u32], mut lo: isize, mut hi: isize) {
    while hi - lo > 16 {
        let mid = lo + (hi - lo) / 2;
        // median-of-three: order a[lo] <= a[mid] <= a[hi]
        if a[mid as usize] < a[lo as usize] {
            a.swap(mid as usize, lo as usize);
        }
        if a[hi as usize] < a[lo as usize] {
            a.swap(hi as usize, lo as usize);
        }
        if a[hi as usize] < a[mid as usize] {
            a.swap(hi as usize, mid as usize);
        }
        let pivot = a[mid as usize];

        let (mut i, mut j) = (lo, hi);
        while i <= j {
            while a[i as usize] < pivot {
                i += 1;
            }
            while a[j as usize] > pivot {
                j -= 1;
            }
            if i <= j {
                a.swap(i as usize, j as usize);
                i += 1;
                j -= 1;
            }
        }
        // recurse into the smaller half, loop on the larger: depth stays O(log n)
        if j - lo < hi - i {
            quicksort(a, lo, j);
            lo = i;
        } else {
            quicksort(a, i, hi);
            hi = j;
        }
    }
    insertion_sort(a, lo, hi);
}

fn bench_quicksort(n: i32) -> i64 {
    let n = n as usize;
    let mut rng = Rng::new(12345);
    let mut a: Vec<u32> = (0..n).map(|_| rng.next()).collect();
    quicksort(&mut a, 0, n as isize - 1);
    let mut h: u32 = 0;
    for i in 0..n {
        h = h.wrapping_mul(31).wrapping_add(a[i]); // order-sensitive checksum
    }
    h as i64
}

// ---------- 4. word count: string hashing + hash map ----------
// std::collections::HashMap is one line, like Go's and Java's. It also hashes
// with SipHash by default -- cryptographically strong, and slower than the
// others' hashes. That choice shows up in this number; see the README.
const VOCAB: usize = 5000;

fn bench_wordcount(n: i32) -> i64 {
    let mut rng = Rng::new(12345);
    let mut words: Vec<String> = Vec::with_capacity(VOCAB);
    for _ in 0..VOCAB {
        let len = 3 + (rng.next() % 6) as usize;
        let mut w = String::with_capacity(len);
        for _ in 0..len {
            w.push((b'a' + (rng.next() % 26) as u8) as char);
        }
        words.push(w);
    }

    let mut counts: HashMap<&str, i64> = HashMap::with_capacity(8192);
    let mut maxc: i64 = 0;
    for _ in 0..n {
        let ra = rng.next() % VOCAB as u32;
        let rb = rng.next() % VOCAB as u32;
        let w: &str = &words[((ra * rb) / VOCAB as u32) as usize]; // triangular, so counts vary
        let c = counts.entry(w).or_insert(0);
        *c += 1;
        if *c > maxc {
            maxc = *c;
        }
    }
    counts.len() as i64 * 1000003 + maxc
}

// ---------- 5. binary trees: allocation, pointer chasing, memory management ----------
// Box per node, ownership does the freeing: dropping the tree at the end of
// each loop iteration walks and deallocates every node, so Rust pays its
// teardown right here in the loop, like C's free_tree and Perl's refcounts.
struct Tree {
    l: Option<Box<Tree>>,
    r: Option<Box<Tree>>,
}

fn make_tree(d: i32) -> Box<Tree> {
    if d == 0 {
        Box::new(Tree { l: None, r: None })
    } else {
        Box::new(Tree {
            l: Some(make_tree(d - 1)),
            r: Some(make_tree(d - 1)),
        })
    }
}

fn check_tree(t: &Tree) -> i64 {
    match (&t.l, &t.r) {
        (Some(l), Some(r)) => 1 + check_tree(l) + check_tree(r),
        _ => 1,
    }
}

fn bench_binarytrees(n: i32) -> i64 {
    let mut h: u32 = 0;
    for k in 0..n {
        let t = make_tree(11); // 4,095 nodes, built and thrown away
        h = h
            .wrapping_mul(31)
            .wrapping_add(check_tree(&t) as u32)
            .wrapping_add(k as u32);
    }
    h as i64
}

// ---------- 6. matmul: triple-nested loops over flat 2D arrays ----------
fn bench_matmul(n: i32) -> i64 {
    let n = n as usize;
    let mut rng = Rng::new(12345);
    let a: Vec<u32> = (0..n * n).map(|_| rng.next() % 100).collect();
    let b: Vec<u32> = (0..n * n).map(|_| rng.next() % 100).collect();
    let mut c = vec![0u32; n * n];
    for i in 0..n {
        let ib = i * n;
        // zip the row of a against b in n-sized rows: chunks_exact tells the
        // optimiser every brow has length n, so with j < n all three bounds
        // checks fold away. Same k order, same wrapping sums, same checksum.
        let arow = &a[ib..ib + n];
        let crow = &mut c[ib..ib + n];
        for j in 0..n {
            let mut s: u32 = 0;
            for (av, brow) in arow.iter().zip(b.chunks_exact(n)) {
                s = s.wrapping_add(av.wrapping_mul(brow[j]));
            }
            crow[j] = s;
        }
    }
    let mut h: u32 = 0;
    for &v in &c {
        h = h.wrapping_mul(31).wrapping_add(v);
    }
    h as i64
}

// ---------- hidden: prng conformance check, not part of the scored suite ----------
// run.py --selftest uses it to validate each language's PRNG directly; several
// languages implement the 64-bit multiply in 32-bit halves and this checks every bit.
fn bench_prng(n: i32) -> i64 {
    let mut rng = Rng::new(12345);
    let mut h: u32 = 0;
    for _ in 0..n {
        h = h.wrapping_mul(31).wrapping_add(rng.next());
    }
    h as i64
}

fn dispatch(name: &str, size: i32) -> i64 {
    match name {
        "mandelbrot" => bench_mandelbrot(size),
        "sieve" => bench_sieve(size),
        "quicksort" => bench_quicksort(size),
        "wordcount" => bench_wordcount(size),
        "binarytrees" => bench_binarytrees(size),
        "matmul" => bench_matmul(size),
        "prng" => bench_prng(size),
        _ => {
            eprintln!("unknown benchmark: {}", name);
            process::exit(2);
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("usage: {} <benchmark> <size> [reps]", args[0]);
        process::exit(2);
    }
    let name = args[1].clone();
    let size: i32 = args[2].parse().unwrap_or(0);
    let mut reps: i32 = if args.len() > 3 {
        args[3].parse().unwrap_or(1)
    } else {
        1
    };
    if reps < 1 {
        reps = 1;
    }

    // Best-of-N: the fastest run is the one least polluted by scheduler noise,
    // cold caches and (for the JIT languages) not-yet-compiled code.
    let mut best = f64::INFINITY;
    let mut result: i64 = 0;
    for r in 0..reps {
        let t0 = Instant::now();
        let v = dispatch(&name, size);
        let elapsed = t0.elapsed().as_secs_f64() * 1000.0;
        if elapsed < best {
            best = elapsed;
        }
        if r == 0 {
            result = v;
        } else if v != result {
            eprintln!("nondeterministic result!");
            process::exit(3);
        }
    }

    println!("OK {} {} {:.3}", name, result, best);
}
